allensdk/brain_observatory/behavior/vbo_2023_nwb_writers/nwb_writer_utils.py

```diff
@@ -0,0 +1,92 @@
+import logging
+from typing import List, Union
+import pytz
+
+from allensdk.brain_observatory.behavior.behavior_ophys_experiment import (
+    BehaviorOphysExperiment,
+)
+from allensdk.brain_observatory.behavior.behavior_session import (
+    BehaviorSession,
+)
+from allensdk.brain_observatory.behavior.data_files.stimulus_file import (
+    BehaviorStimulusFile,
+)
+from allensdk.core.auth_config import LIMS_DB_CREDENTIAL_MAP
+from allensdk.internal.api import db_connection_creator
+
+
+def _compare_metadata(
+    input_session: Union[BehaviorSession, BehaviorOphysExperiment],
+    table_metadata: dict,
+    skip_metadata: List[str],
+    skip_stim: List[str],
+):
+    """Compare data in the metadata table, loaded behavior session, and
+    the BehaviorStimulusFile to each other. Raise on conflicts.
+
+    Parameters
+    ----------
+    input_session : BehaviorSession or BehaviorOphysExperiment
+        Session object we are creating an NWB file for.
+    table_metadata : dict
+        Dictionary of data containing the metadata from the appropriate
+        metadata table.
+    skip_metadata : list of strings
+        Names of metadata keys to skip during comparison.
+    skip_stim : list of strings
+        Names of stimulus file keys to skip during comparison.
+    """
+    try:
+        input_id = input_session.ophys_experiment_id
+        input_type = "BehaviorOphysExperiment"
+    except AttributeError:
+        input_id = input_session.behavior_session_id
+        input_type = "BehaviorSession"
+
+    error_message = ""
+    # Test BehaviorSession object metadata against the metadata from the
+    # behavior session table.
+    bs_metadata = input_session.metadata
+    for key, bs_val in table_metadata.items():
+        if key in skip_metadata:
+            logging.info(f"Skipping metadata table {key} comparison...")
+            continue
+        if bs_val != bs_metadata[key]:
+            error_message += (
+                f"Value for {key} does not match for id={input_id} "
+                f"when comparing {input_type} object metadata and associated "
+                "metadata table.\n"
+                f"\tObject data={bs_metadata[key]};\n"
+                f"\tTable data={bs_val}.\n"
+            )
+
+    # After passing table to object comparison, check object against stimulus
+    # file for final consistency.
+    behavior_session_id = input_session.behavior_session_id
+    db_conn = db_connection_creator(
+        fallback_credentials=LIMS_DB_CREDENTIAL_MAP
+    )
+    stimulus_file = BehaviorStimulusFile.from_lims(
```

Comment:
have you tested this against a full release? Loading stimulus files without multiprocessing can be very slow. There is already `get_session_metadata_multiprocessing` to support multiprocessing

Reply:
Have already produced a full release's worth of VBO data. It ran in tractable time. I'll look into using it through in case some we get handed a larger dataset.